File: src/mvmctl/api/inputs/_network_input.py
"""Network input models for API boundary — existing resource actions."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING

from mvmctl.core._shared import Database
from mvmctl.core.network._repository import NetworkRepository
from mvmctl.core.network._resolver import NetworkResolver
from mvmctl.exceptions import NetworkNotFoundError
from mvmctl.models import NetworkItem

if TYPE_CHECKING:
    pass

__all__ = ["NetworkInput", "NetworkRequest", "ResolvedNetworkInput"]


@dataclass
class NetworkInput:
    """
    Input model for identifying existing networks.

    Used for operations on existing networks (remove, get, inspect, default).
    Provides identifiers (name or id) to resolve the network from DB.
    """

    name: list[str] = field(default_factory=list)
    id: list[str] = field(default_factory=list)

    force: bool | None = None


@dataclass(frozen=True)
class ResolvedNetworkInput:
    """
    Immutable resolved network request — contains resolved NetworkItem records.

    These records are guaranteed to exist in the DB, making them safe to operate on.
    """

    networks: list[NetworkItem]

    force: bool | None = None


class NetworkRequest:
    """
    Resolve network identifiers to DB records and validate.

    Takes NetworkInput (names/ids) and resolves them to NetworkItem records
    using NetworkResolver. Calls ensure_validate() after resolution.
    """

    _result: ResolvedNetworkInput | None = None

    def __init__(
        self, *, inputs: NetworkInput, db: Database | None = None
    ) -> None:
        """Initialize the resolver with database and sub-resolvers."""
        self._inputs = inputs
        self._db = db if db is not None else Database()
        self._network_resolver = NetworkResolver(
            NetworkRepository(self._db),
            include=["leases"],
        )

    @property
    def result(self) -> ResolvedNetworkInput | None:
        return self._result

    def resolve(self) -> ResolvedNetworkInput:
        """
        Resolve network identifiers to NetworkItem records.

        Returns:
            ResolvedNetworkInput with resolved network records.

        Raises:
            NetworkNotFoundError: If any identifier cannot be resolved.

        """
        identifiers = self._inputs.name + self._inputs.id

        if not identifiers:
            raise NetworkNotFoundError("No network identifiers provided")

        result = self._network_resolver.resolve_many(identifiers)

        if result.errors and not result.items:
            raise NetworkNotFoundError(
                f"Could not resolve any networks: {', '.join(result.errors)}"
            )

        self._result = ResolvedNetworkInput(
            networks=result.items,
            force=self._inputs.force if self._inputs.force else False,
        )

        # Validate
        self.ensure_validate()

        return self._result

    def ensure_validate(self) -> None:
        """Validate resolved network inputs."""
        if self._result is None:
            raise NetworkNotFoundError(
                "Failed to resolve necessary dependencies to validate"
            )

        if not self._result.networks:
            raise NetworkNotFoundError("No networks found matching identifiers")

File: pkg/api/inputs/network_input.go
package inputs

import (
	"context"
	"fmt"
	"mvmctl/internal/core/network"
	"mvmctl/internal/lib/model"
	"mvmctl/pkg/errs"
	"strings"
)

// NetworkInput is the raw input for identifying existing networks.
type NetworkInput struct {
	Identifiers []string `json:"identifiers"`
	Force       bool     `json:"force"`
}

// Validate checks that the network input has valid identifiers.
func (i *NetworkInput) Validate() error {
	if len(i.Identifiers) == 0 {
		return fmt.Errorf("at least one network identifier is required")
	}
	return nil
}

// Resolve resolves all identifiers in the input to NetworkItem objects.
// Delegates to network.Resolver.ResolveMany for batch resolution with
// deduplication and error collection.
func (i *NetworkInput) Resolve(ctx context.Context, repo network.Repository) ([]*model.NetworkItem, error) {
	if err := i.Validate(); err != nil {
		return nil, err
	}
	resolver := network.NewResolver(repo, []string{"leases"})
	result, err := resolver.ResolveMany(ctx, i.Identifiers)
	if err != nil {
		return nil, errs.New(errs.CodeNetworkNotFound, err.Error())
	}
	if len(result.Items) == 0 {
		return nil, errs.NotFound(errs.CodeNetworkNotFound, "No networks found matching identifiers")
	}
	if len(result.Errors) > 0 {
		return result.Items, fmt.Errorf("partial resolve failures: %s", strings.Join(result.Errors, "; "))
	}
	return result.Items, nil
}
